src/main/scala/org/zalando/kanadi/api/Subscriptions.scala

```diff
@@ -524,13 +542,197 @@ final case class CancelledByClient(subscriptionId: SubscriptionId, streamId: Str
+                   case Some(futureProvider) =>
+                     futureProvider.value().map { oAuth2Token =>
+                       toHeader(oAuth2Token) +: streamHeaders
+                     }
+                 }
+
+      request = HttpRequest(HttpMethods.GET, uri, headers)
+      _       = logger.debug(request.toString)
+
+      // Create a single connection to avoid the pool
+      connectionFlow = {
+        val host = baseUri_.authority.host.toString()
+        val port = baseUri_.effectivePort
+        if (request.uri.scheme.equalsIgnoreCase("https"))
+          http.outgoingConnectionHttps(host = host, port = port)
+        else http.outgoingConnection(host = host, port = port)
+      }
+
+      response: HttpResponse <- Source
+                                  .single(request)
+                                  .via(connectionFlow)
+                                  .runWith(Sink.head)
+                                  .map(decodeCompressed)
+      _ = logger.debug(response.toString)
+      result <- {
+        if (response.status.isSuccess()) {
+          val streamId = (for {
+            asString <- response.headers.find(_.is(xNakadiStreamIdHeader.toLowerCase))
+          } yield StreamId(asString.value())).getOrElse(
+            throw new ExpectedHeader(xNakadiStreamIdHeader, request, response)
+          )
+
+          val batchParseFlow = Flow.fromFunction[ByteString, Try[ConsumptionBatch]](bs =>
+            Try(ConsumptionBatch.getDecoder.decode(bs.toArray)))
+
+          val graph = response.entity.dataBytes
+            .via(batchParseFlow)
+            .viaMat(KillSwitches.single)(Keep.right)
+            .alsoTo(Sink.onComplete { data =>
+              val cancelledByClient = data match {
+                case util.Failure(CancelledByClient(_, _)) => true
+                case _                                     => false
+              }
+              cleanup(streamId, cancelledByClient)
+              ()
+            })
+            .map { data =>
+              logger.debug(s"SubscriptionId: ${subscriptionId.id.toString}, StreamId: ${streamId.id}")
+              data
+            }
+            .via(parseUserPayloadFlow)
+            .map {
+              case Left(error)   => throw error
+              case Right(result) => result
+            }
+            .withAttributes(ActorAttributes.supervisionStrategy(eventStreamSupervisionDecider
+              .decider(Subscriptions
+                .EventStreamContext(flowId, subscriptionId, streamId, this))))
+
+          Future.successful(Subscriptions.NakadiSource(streamId, graph, request))
+        } else
+          response.status match {
+            case StatusCodes.NotFound =>
+              unmarshalStringOrProblem(response.entity.withContentType(ContentTypes.`application/json`)).map {
+                stringOrProblem =>
+                  throw Subscriptions.Errors.SubscriptionNotFound(request, response, stringOrProblem)
```

Comment:
IMHO, we need more consistency in `Future` / throw.
Otherwise it becomes difficult to follow the code.

Is it worth to fail the future instead of throwing an error?

Reply:
+100